### PR TITLE
Add custom image sample documentation

### PR DESCRIPTION
**Pull request by rzgry**
**What this PR does / why we need it**:
Adds  documentation on how to use a custom image with KFServing

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #673 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
NONE
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/678)

### CODE REVIEW

**Comment by animeshsingh** [comment]
thanks @rzgry - can we investigate if we can also add an advanced version, which is to take a real model from IBM MAX or TF repo,  wrap it and show the procedure?

File: docs/samples/custom/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Predict on a InferenceService using a Custom Image
+
+## Setup
+
+1. Your ~/.kube/config should point to a cluster with [KFServing installed](https://github.com/kubeflow/kfserving/blob/master/docs/DEVELOPER_GUIDE.md#deploy-kfserving).
+2. Your cluster's Istio Ingress gateway must be network accessible.
+
+## Build and push the sample Docker Image
+
+Use Docker to build the sample python server into a container. To build and push with Docker Hub, run these commands replacing {username} with your Docker Hub username:
```
Comment (animeshsingh) [inline docs/samples/custom/README.md]:
```suggestion
```suggestion
Goal of custom image support is to allow users to bring their own wrapped model inside a container and serve it with KFServing. Please note that you will need to ensure that your container is also running a web server e.g. Flask to expose your model endpoints.

In this example we use Docker to build the sample python server into a container. To build and push with Docker Hub, run these commands replacing {username} with your Docker Hub username:
```
```

File: docs/samples/custom/custom.yaml [hunk 2]
```diff
@@ -3,13 +3,13 @@ kind: InferenceService
...
   default:
     predictor:
       custom:
         container:
-          image: seldonio/mock_classifier:1.0
+          image: {username}/custom-sample
           env:
-            - name: PREDICTIVE_UNIT_SERVICE_PORT
-              value: "8080"
+            - name: TARGET
```
Comment (animeshsingh) [inline docs/samples/custom/custom.yaml]: Should it be port?

File: docs/samples/custom/custom.yaml [hunk 3]
```diff
@@ -3,13 +3,13 @@ kind: InferenceService
...
+              value: "Python KFServing Sample"
```
Comment (animeshsingh) [inline docs/samples/custom/custom.yaml]: This should be port number?

**Comment by animeshsingh** [comment]
We can possibly merge this once the changes are taken care of, but would strongly recommend following this up with a more real world model from TF or PyTorch model zoo...

**Comment by yuzisun** [comment]
thanks for taking the step to document custom story! I agree with animesh for a real model serving example and I also suggest creating the example by extending kfserving sdk so they are consistent with other prebuilt model server. I can definitely help you on that.

**Comment by animeshsingh** [comment]
So this can be a three PRs
1. a. Creation and serving of a  simple hello world server container (as you have done)
   b. Real model pulled from a TF/PyTorch/ or any other popular model zoo, show
2. Prebuilt existing container with a model wrapped inside it, and maybe just enable Predict
3. Extend KFServing SDK to create a container compatible with KFServing format, so that all the functionalities of KFServing can be leveraged

File: docs/samples/custom/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# Use the official lightweight Python image.
+# https://hub.docker.com/_/python
+FROM python:3.7-slim
+
+# Copy local code to the container image.
+ENV APP_HOME /app
+WORKDIR $APP_HOME
+COPY . ./
```
Comment (ellis-bigelow) [inline docs/samples/custom/Dockerfile]: Instead of "Copy ." and using dockerignore, can we be a little more surgical?

File: docs/samples/custom/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+# Install production dependencies.
+RUN pip install Flask gunicorn
```
Comment (ellis-bigelow) [inline docs/samples/custom/Dockerfile]: worth a requirements.txt?

File: docs/samples/custom/app.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import os
```
Comment (ellis-bigelow) [inline docs/samples/custom/app.py]: Our custom should probably use the kfserving library with tornado. WDYT?

**Comment by yuzisun** [comment]
@rzgry this can be a good example for kfserving custom https://github.com/SeldonIO/seldon-models/blob/master/pytorch/moviesentiment_roberta/inference/kfserving/model.py

File: docs/samples/custom/kfserving-sdk/model.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+        output = self.model(input_batch)
+
+        percentage = torch.nn.functional.softmax(output, dim=1)[0] * 100
+
+        _, top_5 = torch.topk(output, 5)
+
+        results = {}
+        for idx in top_5[0]:
+            results[self.classes[idx]] = percentage[idx].item()
```
Comment (Tomcli) [inline docs/samples/custom/kfserving-sdk/model.py]: If you converted the softmax output into percentage, can you add a `%` sign to indicate the confident score is in percentage unit?

File: docs/samples/custom/kfserving-sdk/custom.yml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+apiVersion: serving.kubeflow.org/v1alpha2
+kind: InferenceService
+metadata:
+  labels:
+    controller-tools.k8s.io: "1.0"
+  name: kfservingsdksample
```
Comment (Tomcli) [inline docs/samples/custom/kfserving-sdk/custom.yml]: Maybe rename it as kfserving-custom-model?

File: docs/samples/custom/kfserving-sdk/README.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
+# Predict on a InferenceService using the KFServing sdk
```
Comment (Tomcli) [inline docs/samples/custom/kfserving-sdk/README.md]: Maybe rename `KFServing sdk` to `KFServing Model Server`? KFServing sdk also could be the client library which may confuse our users.

File: docs/samples/custom/kfserving-sdk/model.py (hunk 7 above)
Comment (Tomcli) [reply docs/samples/custom/kfserving-sdk/model.py]: Usually for confidence score we don't need to convert it to percentage because confidence score  usually represented in probability.

**Comment by animeshsingh** [comment]
@rzgry @Tomcli can you get together to address the comments which can be done quickly, for the ones which will take longer open an issue which can be the base for a follow-on PR, and lets get this merged

**Comment by rzgry** [comment]
From https://github.com/kubeflow/kfserving/pull/678#issuecomment-586064632 1a and 1b are completed in this PR. 
I can open issues to cover 2 and 3 in later PRs.

**Comment by animeshsingh** [comment]
thanks @rzgry - please open one, assign it to yourself and @Tomcli, and list the issue here in this PR

**Comment by rzgry** [comment]
Opened https://github.com/kubeflow/kfserving/issues/694

**Comment by animeshsingh** [comment]
/lgtm
/approve

Followon to be done via https://github.com/kubeflow/kfserving/issues/694
